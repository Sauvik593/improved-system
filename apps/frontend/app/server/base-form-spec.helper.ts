export const createActionData = async (payload: Record<string, string>) => {
  const params = {};
  const context = {} as unknown as any;
  const formData = new URLSearchParams();

  Object.entries(payload).forEach(([key, value]) => {
    formData.append(key, value);
  });

  const request = new Request('http://new-frontend.kyero.test/', {
    method: 'POST',
    body: formData,
  });

  return { request, params, context };
};
