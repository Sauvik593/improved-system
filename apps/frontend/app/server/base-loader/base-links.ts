import { type LinkDescriptor } from '@remix-run/node';
import { assetsPathTo } from '~/common/client-router/helpers';

export class BaseLinks {
  public get links(): LinkDescriptor[] {
    return [...this.applicationLinks];
  }

  public getDynamicLinks(loader: unknown): LinkDescriptor[] {
    return [];
  }

  private get applicationLinks() {
    return [
      {
        rel: 'apple-touch-icon',
        sizes: '180x180',
        href: assetsPathTo('/favicon/apple-touch-icon.png'),
      },
      {
        rel: 'icon',
        type: 'image/png',
        href: assetsPathTo('/favicon/favicon-32x32.png'),
        sizes: '32x32',
      },
      {
        rel: 'icon',
        type: 'image/png',
        href: assetsPathTo('/favicon/favicon-16x16.png'),
        sizes: '16x16',
      },
      { rel: 'mask-icon', href: assetsPathTo('/favicon/safari-pinned-tab.svg'), color: '#1F4DEF' },
      { rel: 'shortcut icon', href: assetsPathTo('/favicon/favicon.ico') },
      { rel: 'manifest', href: '/manifest.json' },
      {
        rel: 'preload',
        href: assetsPathTo('/fonts/gilroy-medium-webfont.woff2'),
        as: 'font',
        crossOrigin: 'anonymous',
      } as LinkDescriptor,
      {
        rel: 'preload',
        href: assetsPathTo('/fonts/gilroy-bold-webfont.woff2'),
        as: 'font',
        crossOrigin: 'anonymous',
      } as LinkDescriptor,
      {
        rel: 'preload',
        href: assetsPathTo('/fonts/gilroy-semibold-webfont.woff2'),
        as: 'font',
        crossOrigin: 'anonymous',
      } as LinkDescriptor,
    ];
  }
}
