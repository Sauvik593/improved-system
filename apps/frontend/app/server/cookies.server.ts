export const getCookies = (request: Request) => request.headers.get('Cookie');

export const parseSingleCookie = (cookieName: string) => (cookieString: string) =>
  cookieString
    .split(';')
    .find((cookie: string) => cookie.includes(cookieName))
    ?.split('=')[1];

export const getExternalCookie = (cookieName: string) => (cookieString: string | null) =>
  cookieString ? parseSingleCookie(cookieName)(cookieString) : null;

export const getNonEssentialCookie = getExternalCookie('non_essential_cookies');
