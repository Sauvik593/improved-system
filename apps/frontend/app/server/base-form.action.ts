import { json, type ActionArgs, redirect } from '@remix-run/node';
import { validationError } from 'remix-validated-form';
import { withZod } from '@remix-validated-form/with-zod';

import { type TFunction } from 'i18next';
import { type ZodType } from 'zod';
import { type ApiServiceResponseType } from './api/service.server';

import i18next from '~/i18next.server';

import { ApiError, type ApiErrorMessage } from './api/errors.server';
import { LocaleFromPath } from '~/common/i18n/locale-from-path.server';
import { passCookieHeaderFromAPI } from './helpers.server';

export class BaseFormAction<T extends ZodType, E extends {}> {
  request: ActionArgs['request'];
  params: ActionArgs['params'];
  context: ActionArgs['context'];

  constructor(args: ActionArgs) {
    this.request = args.request;
    this.params = args.params;
    this.context = args.context;
  }

  async action() {
    const locale = (await LocaleFromPath.get(this.request)) || 'en';

    const { data, error } = await this.validate(locale);

    if (error) {
      return validationError(error);
    }

    const preparedPayload = this.preparePayload(data);

    try {
      const [apiData, resp, status] = await this.apiCall(preparedPayload, locale);

      if (status === 'error') {
        return this.handleApiErrorMessage(apiData);
      }

      // @ts-ignore
      if (resp.status === 302 && apiData?.to) {
        // @ts-ignore
        return redirect(apiData.to, {
          headers: passCookieHeaderFromAPI(resp),
        });
      }

      // @ts-ignore
      return json(this.prepareSuccessResponse(apiData, resp, preparedPayload), {
        headers: passCookieHeaderFromAPI(resp),
      });
    } catch (error) {
      if (error instanceof Error) {
        return json({
          state: 'error',
          error: error instanceof ApiError ? error.message : 'Unknown error',
        });
      }
    }
  }

  async validate(locale: string) {
    const t = await i18next.getFixedT(locale);
    const schema = this.getValidationSchema(t);
    const { data, error } = await withZod(schema).validate(await this.request.formData());

    return { data, error };
  }

  getValidationSchema = (t: TFunction): T => {
    throw new Error('Not implemented');
  };

  handleApiErrorMessage = ({ errors }: ApiErrorMessage) => {
    const { base, ...restErrors } = errors;
    if (base) {
      return json({ state: 'error', error: base[0] });
    }

    // @ts-ignore
    return validationError({
      // @ts-ignore
      fieldErrors: restErrors,
    });
  };

  preparePayload(payload: T['_output']) {
    return payload;
  }

  apiCall(payload: T['_output'], locale: string): ApiServiceResponseType<E> {
    throw new Error('Not implemented API action call');
  }

  prepareSuccessResponse(data: E, resp: Response, payload: T['_output']) {
    return data;
  }
}
