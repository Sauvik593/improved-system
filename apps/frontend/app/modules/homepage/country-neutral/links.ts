import { BaseLinks } from '~/server/base-loader/base-links';
import { type LinkDescriptor } from '@remix-run/node';

import { SUPPORTED_LOCALES, FALLBACK_LOCALE } from '~/i18n';
import { type CountryNeutralHomepageLoaderType } from './loader.server';
import { type RouteLinks } from '~/server/base-loader/base-loader.server';
import { urlOf } from '~/common/client-router/helpers';

export class CountryNeutralHomepageLinks extends BaseLinks {
  private routeLinks: RouteLinks | undefined;

  public getDynamicLinks({
    locale,
    routeLinks,
  }: CountryNeutralHomepageLoaderType): LinkDescriptor[] {
    this.routeLinks = routeLinks;

    return [this.canonicalLink(locale), ...this.getAlternateLinks(), this.xDefaultLink(locale)];
  }

  private getAlternateLinks() {
    return [...SUPPORTED_LOCALES.map(this.alternateLink)];
  }

  private alternateLink = (locale: string) => {
    const routeLinks = this.routeLinks as RouteLinks;

    return {
      rel: 'alternate' as const,
      href: urlOf(routeLinks.internal.homepage),
      hrefLang: locale,
    };
  };

  private xDefaultLink = (locale: string) => ({
    ...this.alternateLink(FALLBACK_LOCALE),
    hrefLang: 'x-default',
  });

  private canonicalLink = (locale: string) => ({
    ...this.alternateLink(locale),
    rel: 'canonical' as const,
  });
}
