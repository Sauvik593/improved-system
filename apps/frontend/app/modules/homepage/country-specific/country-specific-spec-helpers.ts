import { vitest, beforeAll, type MockedFunction } from 'vitest';

import { type CountryDecorator } from './country.decorator.server';
import {
  CountryHomepageLoader,
  type CountryLoaderResponse,
  type CountryLoaderData,
} from './country.loader.server';
import { CountryHomepageMeta } from './country.meta.server';

import { type ServerCountry } from './helpers';
import { createContext, SERVER_ENV } from 'test/__mocks__/server-env.mock';
import { type LoaderResponse } from '~/server/base-loader/base-loader.server';
import { getCountryRegions } from '~/modules/homepage/api/get-country-regions.server';

vitest.mock('~/modules/homepage/api/get-country-regions.server', () => ({
  getCountryRegions: vitest.fn().mockResolvedValue([]),
}));

const mockedGetCountryRegions = getCountryRegions as MockedFunction<typeof getCountryRegions>;

export const runCountryLoaderSpec = (
  name: string,
  Loader: typeof CountryHomepageLoader,
  Decorator: typeof CountryDecorator,
  country: ServerCountry,
) => {
  const request = new Request('/');
  const params = {};
  const context = createContext();

  const loaderArgs = { request, params, context };

  // @ts-ignore
  const mockedFetch: FetchMock = global.fetch;

  describe(name, () => {
    beforeAll(() => {
      mockedGetCountryRegions.mockResolvedValue([]);
      mockedFetch.mockResponse(JSON.stringify({}));
    });

    afterAll(() => {
      mockedGetCountryRegions.mockReset();
      mockedFetch.mockReset();
    });

    it('should inherit from BaseLoader', () => {
      expect(new Loader(loaderArgs)).toBeInstanceOf(CountryHomepageLoader);
    });

    describe('#init', async () => {
      let data: LoaderResponse<CountryLoaderData>['data'];

      beforeAll(async () => {
        const loader = await new Loader(loaderArgs).init();
        const resp = await loader.json();

        data = resp.data;
      });

      it('should have a correct country returned from loader', async () => {
        expect(data.country).toEqual(country);
        expect(data.country).toMatchObject<ServerCountry>;
      });

      it('should have a correct CountryDecorator', async () => {
        expect(data.ui).toMatchObject<typeof Decorator>;
      });
    });

    describe('#getCacheKey', () => {
      it('should return correct cache key', () => {
        const loader = new Loader(loaderArgs);
        const cacheKey = loader.getCacheKey('en');

        expect(cacheKey).toEqual(`FRONTEND.development.en.${country.id}.homepage.v.0.8`);
      });
    });

    describe('#getCacheExpireAt', () => {
      beforeEach(() => {
        vitest.useFakeTimers();
      });

      afterEach(() => {
        vitest.useRealTimers();
      });

      it('should return correct cache expire time next day at midnight', () => {
        // Set fake timer for the date
        const date = new Date(2022, 1, 1, 10);
        vitest.setSystemTime(date);

        const loader = new Loader(loaderArgs);
        const expireAt = loader.getCacheExpireAt();

        // Timestamp of 2022-02-01 00:00:00
        expect(expireAt).toEqual(1643760000000);
      });
    });
  });
};

const t = vitest.fn().mockImplementation((t: string) => t);
const locale = 'en';

const baseI18n = {
  locale,
  t,
};

export const runCountryMetaTest = (
  name: string,
  countryKey: string,
  CountryMeta: typeof CountryHomepageMeta,
) => {
  describe(name, () => {
    beforeAll(() => {
      mockedGetCountryRegions.mockResolvedValue([]);
    });

    afterEach(() => {
      mockedGetCountryRegions.mockReset();
    });

    it('should inherit from BaseLoader', () => {
      expect(new CountryMeta(baseI18n, SERVER_ENV)).toBeInstanceOf(CountryHomepageMeta);
    });

    it('should have correct title and description', () => {
      const meta = new CountryMeta(baseI18n, SERVER_ENV);
      const metaTags = meta.getAll();

      expect(metaTags[0]).toEqual({
        title: `common.homepage.${countryKey}.meta.title | Kyero`,
      });
      expect(metaTags[1]).toEqual({
        name: 'description',
        content: `common.homepage.${countryKey}.meta.description`,
      });
    });
  });
};

// @ts-ignore
export const createLoader = async (countryLoader: any) => {
  const request = new Request('/');
  const params = {};
  const context = createContext();

  const loaderArgs = { request, params, context };

  const loader = new countryLoader(loaderArgs);
  // @ts-ignore
  const loaderData = await loader.init();
  const data = (await loaderData.json()) as CountryLoaderResponse;

  return data;
};
