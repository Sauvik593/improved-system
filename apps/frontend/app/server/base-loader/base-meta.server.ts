import { FacebookSocialCard } from './social-card/facebook-social-card';
import { TwitterSocialCard } from './social-card/twitter-social-card';

import { type BaseI18n } from './base-loader.server';
import { type V2_HtmlMetaDescriptor } from '@remix-run/node';

import { type SERVER_ENV_SCHEMA } from '~/../env.server';
import { assetsPathTo } from '~/common/client-router/helpers';

type SERVER_ENV_TYPE = ReturnType<typeof SERVER_ENV_SCHEMA.parse>;

export class BaseMeta {
  i18n: BaseI18n;
  SERVER_ENV: SERVER_ENV_TYPE;

  constructor(i18n: BaseI18n, SERVER_ENV: SERVER_ENV_TYPE) {
    this.i18n = i18n;
    this.SERVER_ENV = SERVER_ENV;
  }

  getAll() {
    return [
      ...this.titleMeta,
      ...this.descriptionMeta,
      ...this.robots,
      ...this.owner,
      ...this.copyright,
      ...this.noTranslate,
      ...this.socialMeta,
      ...this.applicationMeta,
      ...this.schema,
    ];
  }

  private get titleMeta() {
    return [{ title: `${this.title} | Kyero` }];
  }

  protected get schema(): V2_HtmlMetaDescriptor[] {
    return [];
  }

  private get descriptionMeta() {
    return [
      {
        name: 'description',
        content: this.description,
      },
    ];
  }

  private get owner() {
    return [
      {
        name: 'owner',
        content: 'Kyero.com',
      },
    ];
  }

  private get copyright() {
    return [
      {
        name: 'copyright',
        content: this.i18n.t('common.meta.copyright', { year: new Date().getFullYear() }),
      },
    ];
  }

  private get noTranslate() {
    return [
      {
        name: 'google',
        content: 'notranslate',
      },
    ];
  }

  protected get title() {
    return this.i18n.t('common.homepage.spain.meta.title');
  }

  private get robots() {
    return [
      {
        name: 'robots',
        content: this.SERVER_ENV.KYERO_ENV === 'production' ? 'index, follow' : 'noindex, nofollow',
      },
    ];
  }

  protected get description() {
    return this.i18n.t('common.homepage.spain.meta.description');
  }

  private get applicationMeta() {
    return [
      { name: 'apple-mobile-web-app-title', content: 'Kyero' },
      { name: 'application-name', content: 'Kyero' },
      { name: 'msapplication-config', content: assetsPathTo('/favicon/browserconfig.xml') },
      { name: 'theme-color', content: '#1F4DEF' },
    ];
  }

  private get socialMeta() {
    return [
      ...new FacebookSocialCard(this.card, this.i18n.locale, this.SERVER_ENV).metaTags,
      ...new TwitterSocialCard(this.card, this.i18n.locale, this.SERVER_ENV).metaTags,
    ];
  }

  protected get card() {
    return {
      title: this.title,
      description: this.description,
      url: this.SERVER_ENV.BASE_URL,
      type: 'article' as const,
      image: undefined,
    };
  }
}
