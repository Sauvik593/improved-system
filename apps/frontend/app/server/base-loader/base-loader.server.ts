import i18next from '~/i18next.server';
import { BaseMeta } from './base-meta.server';
import { LocaleFromPath } from '~/common/i18n/locale-from-path.server';
import { FALLBACK_LOCALE, config } from '~/i18n';

import { type DataFunctionArgs } from '@remix-run/server-runtime';
import { type TFunction } from 'i18next';
import { type V2_HtmlMetaDescriptor, json, type TypedResponse } from '@remix-run/node';

import { RedisCache } from '../redis-cache.server';
import { ClientRouterServer } from '~/common/client-router/client-router.server';

import { COUNTRIES, type ServerCountry } from '~/modules/homepage/country-specific/helpers';
import { handleGenericError, handleNotFound } from '~/modules/errors/helpers.server';
import { getAppRoutes, getCountryLocalisedRoutes } from '../routes-helper.server';

export type BaseI18n = {
  locale: string;
  t: TFunction;
};

export type RouteLinks = ReturnType<typeof BaseLoader.prototype.getRoutes>;

export interface LoaderResponse<T> {
  locale: string;
  localizedRoutes: { locale: string; path: string }[];
  meta: V2_HtmlMetaDescriptor[];
  data: T;
  routeLinks: RouteLinks;
}

export class BaseLoader<T> {
  request: DataFunctionArgs['request'];
  params: DataFunctionArgs['params'];
  context: DataFunctionArgs['context'];
  clientRouter: typeof ClientRouterServer;
  i18n: BaseI18n | undefined;

  constructor(payload: DataFunctionArgs) {
    this.request = payload.request;
    this.params = payload.params;
    this.context = payload.context;
    this.clientRouter = ClientRouterServer;
  }

  async init(status: number | undefined = 200): Promise<TypedResponse<LoaderResponse<T>>> {
    try {
      await this.checkConstraint();

      const locale = (await LocaleFromPath.get(this.request)) || FALLBACK_LOCALE;
      const t = await i18next.getFixedT(locale);

      if (!t) {
        throw new Error('i18n not found');
      }

      this.i18n = {
        locale,
        t,
      };

      const CACHE_KEY = this.getCacheKey(locale);
      const data = CACHE_KEY
        ? await RedisCache.get(CACHE_KEY as string, () => this.getData(), this.getCacheExpireAt())
        : await this.getData();

      return json(data as LoaderResponse<T>, { status });
    } catch (e: Error | Response | unknown) {
      return await handleGenericError(e)({
        context: this.context,
        params: this.params,
        request: this.request,
      });
    }
  }

  async getData() {
    const locale = this.i18n?.locale as string;
    const data = await this.data();

    return {
      locale,
      localizedRoutes: this.localizedRoutes(),
      data: data,
      meta: this.meta(),
      routeLinks: this.getRoutes(locale),
    };
  }

  getCacheKey(locale: string): string | undefined {
    return undefined;
  }

  getCacheExpireAt(): number | undefined {
    return undefined;
  }

  protected async checkConstraint() {
    if (
      !!this.params.locale &&
      !config.supportedLngs.includes(this.params.locale as unknown as any)
    ) {
      return await handleNotFound({
        params: this.params,
        context: this.context,
        request: this.request,
      });
    }

    Promise.resolve();
  }

  async data(): Promise<T> {
    return [] as T;
  }

  protected get country(): ServerCountry {
    return COUNTRIES.SPAIN;
  }

  protected localizedRoutes(): { locale: string; path: string }[] {
    return getCountryLocalisedRoutes(this.country);
  }

  getRoutes(locale: string) {
    return getAppRoutes({ context: this.context, locale, country: this.country });
  }

  protected meta() {
    if (!this.i18n) {
      throw new Error('i18n not found');
    }

    return new this.MetaClass(this.i18n as BaseI18n, this.context.SERVER_ENV).getAll();
  }

  protected get MetaClass() {
    return BaseMeta;
  }
}
