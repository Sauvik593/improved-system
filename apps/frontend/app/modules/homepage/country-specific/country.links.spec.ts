import { vitest } from 'vitest';

import { BaseLinks } from '~/server/base-loader/base-links';
import { CountryLinks } from './country.links';

import { PortugalHomepageLoader } from './portugal.server';
import { SpainHomepageLoader } from './spain.server';
import { ItalyHomepageLoader } from './italy.server';
import { FranceHomepageLoader } from './france.server';

import { createLoader } from './country-specific-spec-helpers';

vitest.mock('~/modules/homepage/api/get-country-regions.server', () => ({
  getCountryRegions: vitest.fn().mockResolvedValue([]),
}));

describe('CountryLinks', () => {
  // @ts-ignore
  const mockedFetch: FetchMock = global.fetch;

  beforeAll(() => {
    mockedFetch.mockResponse(JSON.stringify({}));
  });

  afterAll(() => {
    mockedFetch.mockReset();
  });

  it('descends from BaseLinks', () => {
    expect(CountryLinks.prototype).toBeInstanceOf(BaseLinks);
  });

  describe('#getDynamicLinks', () => {
    [
      [PortugalHomepageLoader, 'portugal'],
      [SpainHomepageLoader, 'spain'],
      [ItalyHomepageLoader, 'italy'],
      [FranceHomepageLoader, 'france'],
    ].forEach(([countryLoader, name]) => {
      it(`generates correct locale for ${name}`, async () => {
        const loader = await createLoader(countryLoader);
        const countryLinks = new CountryLinks();
        const dynamicLinks = countryLinks.getDynamicLinks(loader);

        expect(dynamicLinks).toMatchSnapshot();
      });
    });
  });
});
