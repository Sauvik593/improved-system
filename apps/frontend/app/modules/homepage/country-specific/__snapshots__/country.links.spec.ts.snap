// Vitest Snapshot v1

exports[`CountryLinks > #getDynamicLinks > generates correct locale for france 1`] = `
[
  {
    "href": "http://new-frontend.kyero.test/en/france",
    "hrefLang": "en",
    "rel": "canonical",
  },
  {
    "href": "http://new-frontend.kyero.test/ca/fran%C3%A7a",
    "hrefLang": "ca",
    "rel": "alternate",
  },
  {
    "href": "http://new-frontend.kyero.test/da/frankrig",
    "hrefLang": "da",
    "rel": "alternate",
  },
  {
    "href": "http://new-frontend.kyero.test/de/frankreich",
    "hrefLang": "de",
    "rel": "alternate",
  },
  {
    "href": "http://new-frontend.kyero.test/en/france",
    "hrefLang": "en",
    "rel": "alternate",
  },
  {
    "href": "http://new-frontend.kyero.test/es/francia",
    "hrefLang": "es",
    "rel": "alternate",
  },
  {
    "href": "http://new-frontend.kyero.test/fi/ranska",
    "hrefLang": "fi",
    "rel": "alternate",
  },
  {
    "href": "http://new-frontend.kyero.test/fr/france",
    "hrefLang": "fr",
    "rel": "alternate",
  },
  {
    "href": "http://new-frontend.kyero.test/it/francia",
    "hrefLang": "it",
    "rel": "alternate",
  },
  {
    "href": "http://new-frontend.kyero.test/nl/frankrijk",
    "hrefLang": "nl",
    "rel": "alternate",
  },
  {
    "href": "http://new-frontend.kyero.test/no/frankrike",
    "hrefLang": "no",
    "rel": "alternate",
  },
  {
    "href": "http://new-frontend.kyero.test/pt/franca",
    "hrefLang": "pt",
    "rel": "alternate",
  },
  {
    "href": "http://new-frontend.kyero.test/ru/%D1%84%D1%80%D0%B0%D0%BD%D1%86%D0%B8%D1%8F",
    "hrefLang": "ru",
    "rel": "alternate",
  },
  {
    "href": "http://new-frontend.kyero.test/sv/frankrike",
    "hrefLang": "sv",
    "rel": "alternate",
  },
  {
    "as": "image",
    "href": "/new-frontend-assets/images/homepage/v1/france/home-mobile.webp",
    "imageSrcSet": "/new-frontend-assets/images/homepage/v1/france/home-mobile@2x.webp 2x",
    "media": "(max-width: 799px)",
    "rel": "preload",
    "type": "image/webp",
  },
  {
    "as": "image",
    "href": "/new-frontend-assets/images/homepage/v1/france/home-tablet.webp",
    "imageSrcSet": "/new-frontend-assets/images/homepage/v1/france/home-tablet@2x.webp 2x",
    "media": "(min-width: 799.1px) and (max-width: 1279px)",
    "rel": "preload",
    "type": "image/webp",
  },
  {
    "as": "image",
    "href": "/new-frontend-assets/images/homepage/v1/france/home-md.webp",
    "imageSrcSet": "/new-frontend-assets/images/homepage/v1/france/home-md@2x.webp 2x",
    "media": "(min-width: 1279.1px) and (max-width: 1400px)",
    "rel": "preload",
    "type": "image/webp",
  },
  {
    "as": "image",
    "href": "/new-frontend-assets/images/homepage/v1/france/home-hi.webp",
    "imageSrcSet": "/new-frontend-assets/images/homepage/v1/france/home-hi@2x.webp 2x",
    "media": "(min-width: 1400.1px)",
    "rel": "preload",
    "type": "image/webp",
  },
  {
    "href": "http://new-frontend.kyero.test/en/france",
    "hrefLang": "x-default",
    "rel": "alternate",
  },
]
`;

exports[`CountryLinks > #getDynamicLinks > generates correct locale for italy 1`] = `
[
  {
    "href": "http://new-frontend.kyero.test/en/italy",
    "hrefLang": "en",
    "rel": "canonical",
  },
  {
    "href": "http://new-frontend.kyero.test/ca/it%C3%A0lia",
    "hrefLang": "ca",
    "rel": "alternate",
  },
  {
    "href": "http://new-frontend.kyero.test/da/italien",
    "hrefLang": "da",
    "rel": "alternate",
  },
  {
    "href": "http://new-frontend.kyero.test/de/italien",
    "hrefLang": "de",
    "rel": "alternate",
  },
  {
    "href": "http://new-frontend.kyero.test/en/italy",
    "hrefLang": "en",
    "rel": "alternate",
  },
  {
    "href": "http://new-frontend.kyero.test/es/italia",
    "hrefLang": "es",
    "rel": "alternate",
  },
  {
    "href": "http://new-frontend.kyero.test/fi/italia",
    "hrefLang": "fi",
    "rel": "alternate",
  },
  {
    "href": "http://new-frontend.kyero.test/fr/italie",
    "hrefLang": "fr",
    "rel": "alternate",
  },
  {
    "href": "http://new-frontend.kyero.test/it/italia",
    "hrefLang": "it",
    "rel": "alternate",
  },
  {
    "href": "http://new-frontend.kyero.test/nl/itali%C3%AB",
    "hrefLang": "nl",
    "rel": "alternate",
  },
  {
    "href": "http://new-frontend.kyero.test/no/italia",
    "hrefLang": "no",
    "rel": "alternate",
  },
  {
    "href": "http://new-frontend.kyero.test/pt/it%C3%A1lia",
    "hrefLang": "pt",
    "rel": "alternate",
  },
  {
    "href": "http://new-frontend.kyero.test/ru/%D0%B8%D1%82%D0%B0%D0%BB%D0%B8%D1%8F",
    "hrefLang": "ru",
    "rel": "alternate",
  },
  {
    "href": "http://new-frontend.kyero.test/sv/italien",
    "hrefLang": "sv",
    "rel": "alternate",
  },
  {
    "as": "image",
    "href": "/new-frontend-assets/images/homepage/v1/italy/home-mobile.webp",
    "imageSrcSet": "/new-frontend-assets/images/homepage/v1/italy/home-mobile@2x.webp 2x",
    "media": "(max-width: 799px)",
    "rel": "preload",
    "type": "image/webp",
  },
  {
    "as": "image",
    "href": "/new-frontend-assets/images/homepage/v1/italy/home-tablet.webp",
    "imageSrcSet": "/new-frontend-assets/images/homepage/v1/italy/home-tablet@2x.webp 2x",
    "media": "(min-width: 799.1px) and (max-width: 1279px)",
    "rel": "preload",
    "type": "image/webp",
  },
  {
    "as": "image",
    "href": "/new-frontend-assets/images/homepage/v1/italy/home-md.webp",
    "imageSrcSet": "/new-frontend-assets/images/homepage/v1/italy/home-md@2x.webp 2x",
    "media": "(min-width: 1279.1px) and (max-width: 1400px)",
    "rel": "preload",
    "type": "image/webp",
  },
  {
    "as": "image",
    "href": "/new-frontend-assets/images/homepage/v1/italy/home-hi.webp",
    "imageSrcSet": "/new-frontend-assets/images/homepage/v1/italy/home-hi@2x.webp 2x",
    "media": "(min-width: 1400.1px)",
    "rel": "preload",
    "type": "image/webp",
  },
  {
    "href": "http://new-frontend.kyero.test/en/italy",
    "hrefLang": "x-default",
    "rel": "alternate",
  },
]
`;

exports[`CountryLinks > #getDynamicLinks > generates correct locale for portugal 1`] = `
[
  {
    "href": "http://new-frontend.kyero.test/en/portugal",
    "hrefLang": "en",
    "rel": "canonical",
  },
  {
    "href": "http://new-frontend.kyero.test/ca/portugal",
    "hrefLang": "ca",
    "rel": "alternate",
  },
  {
    "href": "http://new-frontend.kyero.test/da/portugal",
    "hrefLang": "da",
    "rel": "alternate",
  },
  {
    "href": "http://new-frontend.kyero.test/de/portugal",
    "hrefLang": "de",
    "rel": "alternate",
  },
  {
    "href": "http://new-frontend.kyero.test/en/portugal",
    "hrefLang": "en",
    "rel": "alternate",
  },
  {
    "href": "http://new-frontend.kyero.test/es/portugal",
    "hrefLang": "es",
    "rel": "alternate",
  },
  {
    "href": "http://new-frontend.kyero.test/fi/portugali",
    "hrefLang": "fi",
    "rel": "alternate",
  },
  {
    "href": "http://new-frontend.kyero.test/fr/portugal",
    "hrefLang": "fr",
    "rel": "alternate",
  },
  {
    "href": "http://new-frontend.kyero.test/it/portogallo",
    "hrefLang": "it",
    "rel": "alternate",
  },
  {
    "href": "http://new-frontend.kyero.test/nl/portugal",
    "hrefLang": "nl",
    "rel": "alternate",
  },
  {
    "href": "http://new-frontend.kyero.test/no/portugal",
    "hrefLang": "no",
    "rel": "alternate",
  },
  {
    "href": "http://new-frontend.kyero.test/pt/portugal",
    "hrefLang": "pt",
    "rel": "alternate",
  },
  {
    "href": "http://new-frontend.kyero.test/ru/%D0%BF%D0%BE%D1%80%D1%82%D1%83%D0%B3%D0%B0%D0%BB%D0%B8%D1%8F",
    "hrefLang": "ru",
    "rel": "alternate",
  },
  {
    "href": "http://new-frontend.kyero.test/sv/portugal",
    "hrefLang": "sv",
    "rel": "alternate",
  },
  {
    "as": "image",
    "href": "/new-frontend-assets/images/homepage/v1/portugal/home-mobile.webp",
    "imageSrcSet": "/new-frontend-assets/images/homepage/v1/portugal/home-mobile@2x.webp 2x",
    "media": "(max-width: 799px)",
    "rel": "preload",
    "type": "image/webp",
  },
  {
    "as": "image",
    "href": "/new-frontend-assets/images/homepage/v1/portugal/home-tablet.webp",
    "imageSrcSet": "/new-frontend-assets/images/homepage/v1/portugal/home-tablet@2x.webp 2x",
    "media": "(min-width: 799.1px) and (max-width: 1279px)",
    "rel": "preload",
    "type": "image/webp",
  },
  {
    "as": "image",
    "href": "/new-frontend-assets/images/homepage/v1/portugal/home-md.webp",
    "imageSrcSet": "/new-frontend-assets/images/homepage/v1/portugal/home-md@2x.webp 2x",
    "media": "(min-width: 1279.1px) and (max-width: 1400px)",
    "rel": "preload",
    "type": "image/webp",
  },
  {
    "as": "image",
    "href": "/new-frontend-assets/images/homepage/v1/portugal/home-hi.webp",
    "imageSrcSet": "/new-frontend-assets/images/homepage/v1/portugal/home-hi@2x.webp 2x",
    "media": "(min-width: 1400.1px)",
    "rel": "preload",
    "type": "image/webp",
  },
  {
    "href": "http://new-frontend.kyero.test/en/portugal",
    "hrefLang": "x-default",
    "rel": "alternate",
  },
]
`;

exports[`CountryLinks > #getDynamicLinks > generates correct locale for spain 1`] = `
[
  {
    "href": "http://new-frontend.kyero.test",
    "hrefLang": "en",
    "rel": "canonical",
  },
  {
    "href": "http://new-frontend.kyero.test/ca",
    "hrefLang": "ca",
    "rel": "alternate",
  },
  {
    "href": "http://new-frontend.kyero.test/da",
    "hrefLang": "da",
    "rel": "alternate",
  },
  {
    "href": "http://new-frontend.kyero.test/de",
    "hrefLang": "de",
    "rel": "alternate",
  },
  {
    "href": "http://new-frontend.kyero.test",
    "hrefLang": "en",
    "rel": "alternate",
  },
  {
    "href": "http://new-frontend.kyero.test/es",
    "hrefLang": "es",
    "rel": "alternate",
  },
  {
    "href": "http://new-frontend.kyero.test/fi",
    "hrefLang": "fi",
    "rel": "alternate",
  },
  {
    "href": "http://new-frontend.kyero.test/fr",
    "hrefLang": "fr",
    "rel": "alternate",
  },
  {
    "href": "http://new-frontend.kyero.test/it",
    "hrefLang": "it",
    "rel": "alternate",
  },
  {
    "href": "http://new-frontend.kyero.test/nl",
    "hrefLang": "nl",
    "rel": "alternate",
  },
  {
    "href": "http://new-frontend.kyero.test/no",
    "hrefLang": "no",
    "rel": "alternate",
  },
  {
    "href": "http://new-frontend.kyero.test/pt",
    "hrefLang": "pt",
    "rel": "alternate",
  },
  {
    "href": "http://new-frontend.kyero.test/ru",
    "hrefLang": "ru",
    "rel": "alternate",
  },
  {
    "href": "http://new-frontend.kyero.test/sv",
    "hrefLang": "sv",
    "rel": "alternate",
  },
  {
    "as": "image",
    "href": "/new-frontend-assets/images/homepage/v1/spain/home-mobile.webp",
    "imageSrcSet": "/new-frontend-assets/images/homepage/v1/spain/home-mobile@2x.webp 2x",
    "media": "(max-width: 799px)",
    "rel": "preload",
    "type": "image/webp",
  },
  {
    "as": "image",
    "href": "/new-frontend-assets/images/homepage/v1/spain/home-tablet.webp",
    "imageSrcSet": "/new-frontend-assets/images/homepage/v1/spain/home-tablet@2x.webp 2x",
    "media": "(min-width: 799.1px) and (max-width: 1279px)",
    "rel": "preload",
    "type": "image/webp",
  },
  {
    "as": "image",
    "href": "/new-frontend-assets/images/homepage/v1/spain/home-md.webp",
    "imageSrcSet": "/new-frontend-assets/images/homepage/v1/spain/home-md@2x.webp 2x",
    "media": "(min-width: 1279.1px) and (max-width: 1400px)",
    "rel": "preload",
    "type": "image/webp",
  },
  {
    "as": "image",
    "href": "/new-frontend-assets/images/homepage/v1/spain/home-hi.webp",
    "imageSrcSet": "/new-frontend-assets/images/homepage/v1/spain/home-hi@2x.webp 2x",
    "media": "(min-width: 1400.1px)",
    "rel": "preload",
    "type": "image/webp",
  },
  {
    "href": "http://new-frontend.kyero.test",
    "hrefLang": "x-default",
    "rel": "alternate",
  },
]
`;
